src/node_http_common-inl.h

```diff
@@ -66,6 +68,341 @@ NgHeaders<T>::NgHeaders(Environment* env, v8::Local<v8::Array> headers) {
+  std::hash<std::string>{}(":protocol")
+};
+static const size_t ng_valid_pseudo_headers_size = ng_valid_pseudo_headers.size();
+
+MUST_USE_RESULT inline bool VALIDATE_FOR_ILLEGAL_CONNECTION_SPECIFIC_HEADER(
+        v8::Isolate*& isolate,
+        const std::string_view& header,
+        const size_t& hash,
+        const std::string_view& value) {
+
+  if (ng_forbidden_headers.contains(hash) ||
+      (hash == te_hash && value == "trailers")) {
+    THROW_ERR_HTTP2_INVALID_CONNECTION_HEADERS(isolate,
+        "HTTP/1 Connection specific headers are forbidden: \"%s\"", header);
+    return false;
+  }
+
+  return true;
+}
+
+MUST_USE_RESULT inline bool VALIDATE_PSEUDO_HEADER(v8::Isolate*& isolate,
+                                  const size_t& hash,
+                                  const std::string_view& name,
+                                  const http_headers_type& type) {
+  if (type == http2_request && ng_valid_pseudo_headers.contains(hash)) {
+    return true;
+  }
+
+  if (type == http2_response && hash == status_hash) {
+    return true;
+  }
+
+  THROW_ERR_HTTP2_INVALID_PSEUDOHEADER(isolate, "\"%s\" is an invalid pseudoheader or is used incorrectly", name);
+  return false;
+}
+
+MUST_USE_RESULT inline bool VALIDATE_SINGLES_HEADER(v8::Isolate*& isolate,
+                                    std::unordered_set<size_t>& singles,
+                                    const std::string_view& header,
+                                    const size_t& header_hash) {
+  if (singles.contains(header_hash)) {
+    THROW_ERR_HTTP2_HEADER_SINGLE_VALUE(isolate, "Header field \"%s\" must only have a single value", header);
+    return false;
+  }
+
+  singles.insert(header_hash);
+  return true;
+}
+
+MUST_USE_RESULT inline bool ToString(v8::Isolate*& isolate,
+                                     const v8::Local<v8::Context>& context,
+                                     const v8::Local<v8::Value>& in,
+                                     v8::Local<v8::String>* out) {
+  v8::Local<v8::String> str_;
+  if (in->IsString()) {
+    str_ = in.As<v8::String>();
+  } else if (!in->ToString(context).ToLocal(&str_)) {
+    return false;
+  }
+
+  // TODO: for headers must be true?
+  if (!str_->IsOneByte()) {
+    return false;
+  }
+
+  *out = str_;
+  return true;
+}
+
+MUST_USE_RESULT inline std::unordered_set<size_t> GetSensitiveHeaders(
+    v8::Isolate*& isolate,
+    const v8::Local<v8::Context>& context,
+    const v8::Local<v8::Object>& headers) {
+
+  std::unordered_set<size_t> neverIndex{};
+
+  // TODO: use #define / constants ? -- whats the actual convention to do it in c++?
+  auto kSensitiveHeaders = v8::Symbol::ForApi(isolate,
+      FIXED_ONE_BYTE_STRING(isolate, "nodejs.http2.sensitiveHeaders"));
+
+  // construct sensitive headers index
+  v8::Local<v8::Value> maybeNeverIndex;
+  v8::Local<v8::Value> val;
+  v8::Local<v8::String> header;
+
+  if (headers->Get(context, kSensitiveHeaders).ToLocal(&maybeNeverIndex) &&
```

Comment:
C++ guide in Node.js says that you should handle `invalid` case first and later continue the code, rather than having a huge if statement like this.